### PR TITLE
Coupled variable function energy

### LINKED ISSUE #18694: Create free energy class based on MooseFunctions
## Reason
Using MooseFunctions to compute the free energy, chemical potentials, and their derivatives (with x,y,z being substituted by concentrations) would enable the use of simple polynomial fits and bilinear interpolations for surrogate free energy models. 
In the long run we need something better for systems with more than 3 concentration variables, but this would be a simple start.

## Design
Have a material that takes functions for up to three chemical potentials (with up to three concentrations) and optionally a function for the free energy itself. Derivatives of the chemical potentials would be computed using `Function::gradient()` .

## Impact
Added capability

### PR DESCRIPTION
**Pull request by dschwen**
Closes #18694

Still needs a test.

@parikshitbajpai @tonkmr @chaitanyaBhave26

### CODE REVIEW

**Comment by dschwen** [comment]
image

Grain growth test with pretabulated 4 OP energy, vs. PolyCrystal Kernel Action, after 100 timesteps (three circular grains in a matrix).

**Comment by dschwen** [comment]
Note that this uses finite differencing as a proof of concept. I'll work on adding proper derivatives later.

**Comment by dschwen** [comment]
No module names `numpy` for the table generation. I thought that was fixed a long time ago when I ran into this previously. But maybe that was for an app @milljm

File: modules/phase_field/test/tests/free_energy_material/tests [hunk 1]
```diff
@@ -97,4 +97,23 @@
...
+  [CoupledValueFunctionFreeEnergy]
+    issues = '#18694'
+    design = 'CoupledValueFunctionFreeEnergy.md'
+    requirement = 'The system shall provide an object to compute the a free energy and its chemical '
+                  'potentials for up to four constituents from a set of MooseFunction objects'
+
+    [prepare_tabulation]
+      type = 'RunCommand'
+      command = './tabulate_grain_growth_energy.py'
+      detail = 'With a script for tabulating grain growth energies provided'
```
Comment (GiudGiud) [inline modules/phase_field/test/tests/free_energy_material/tests:110]:
```suggestion
      detail = 'with a script for tabulating grain growth energies provided'
```

File: modules/phase_field/test/tests/free_energy_material/tests [hunk 2]
```diff
@@ -97,4 +97,23 @@
...
+    []
+    [tabulated_grain_growth]
+      type = 'Exodiff'
+      input = 'CoupledValueFunctionFreeEnergy.i'
+      exodiff = 'CoupledValueFunctionFreeEnergy_out.e'
+      detail = 'Using a pretabulated free energy through PiecewiseMultilinear functions'
```
Comment (GiudGiud) [inline modules/phase_field/test/tests/free_energy_material/tests:116]:
```suggestion
      detail = 'using a pre-tabulated free energy through PiecewiseMultilinear functions'
```

**Review by GiudGiud — APPROVED** [review]
finite difference does a pretty good job here ;)
just curious, what are you going to define for the gradient at slope breaks?

**Comment by dschwen** [comment]
> just curious, what are you going to define for the gradient at slope breaks?

Yeah, good question. I was hoping to simply ignore the grid edges and pretend I'm on the < element. Thats effectively what I'm doing with the zero gradient for the constant grid interpolation.

**Comment by dschwen** [comment]
I'll merge this and address the typos as part of another open PR I have. I have some more capitalized details in there

**Review by laagesen — COMMENTED** [review]
Looks like a nice capability @dschwen. I will make sure folks from my group know about it.

**Comment by dschwen** [comment]
> but which element? Lowest Id one kind of convention?

@GiudGiud those are not elements but grid cells in an orthogonal 1D,2D,3D, or 4D grid. I'd go with `<` based on coordinate components in grid space.

**Pull request merged**